Storage: Improve volume metadata

roosterfish opened the pull request:
Spec here https://docs.google.com/document/d/1AKm3Jgt08ovNShrrP27P7AIjbdtTZjc7RZ4fMpCiHK8/edit?tab=t.0

### Overview

This PR improves the storage volume metadata to be able to account for multiple volumes.
This is especially useful in case of instances to allow recovery of attached custom storage volumes.
It adds a new format identified by a new version field which can be set by the most recent CLI when exporting backups using `--export-version=(1|2)`.

Changing the format requires modifications for backup/restore and migration as both make use of the same data structures.
Importing backups from newer LXDs on older LXDs and being able to migrate from newer to older should still work seamlessly.

In addition a new API extension `backup_metadata_version` is added which acts as the gatekeeper for being able to set the version of the backup's format.

### Backport info

Most of the PR can be backported to 5.21, except the last commits indicating that this is only intended for 6.x going forward.
As the PR changes lots of lines, this should make backporting of other features much easier as the internal changes are getting backported whilst the artifacts that LXD writes to file still remain to use the old format (1).

### PR dependencies

PRs to fix the panic in case of invalid backup file format. This panic would currently happen when importing newer backups on older versions of LXD or when modifying the backup artifact:
* https://github.com/canonical/lxd/pull/15165 + https://github.com/canonical/lxd/pull/15168
* https://github.com/canonical/lxd/pull/15163
* https://github.com/canonical/lxd/pull/15164
* https://github.com/canonical/lxd/pull/15166

Connected PR in lxd-ci for tests: https://github.com/canonical/lxd-ci/pull/443

tomponline commented:
Can we get b3a889348666546f13be1763af27eb6edef7b1db as a separate PR please

roosterfish commented:
Ready now for a first review.

minaelee commented on doc/api-extensions.md:
```diff
@@ -2630,3 +2630,8 @@ Adds the field `client_certificate` to `GET /1.0` to indicate if the current req
+## `backup_version`
+
+Adds the field `version` when exporting instances and custom storage volumes to define the backup file format.
+In case the field is omitted the original (old) version is used to allow staying compatible with old clients.
```
```suggestion
In case the field is omitted, the original (old) version is used to maintain compatibility with older clients.
```

roosterfish replied:
Thanks, fixed.

minaelee commented on doc/howto/storage_backup_volume.md:
```diff
@@ -139,6 +139,11 @@ You can add any of the following flags to the command:
   Snapshots are exported as differences from the main volume, which decreases their size (quota) and makes them easily accessible.
+
+`--version`
+: If you intend to import the backup on an older version of LXD, you can set the version to `1.0` which will use the original (old)
```
```suggestion
: If you intend to import the backup to an older version of LXD, set the version to `1.0` which will use the original (old)
```

roosterfish replied:
Fixed.

minaelee commented on doc/howto/storage_backup_volume.md:
```diff
@@ -139,6 +139,11 @@ You can add any of the following flags to the command:
+`--version`
+: If you intend to import the backup on an older version of LXD, you can set the version to `1.0` which will use the original (old)
+backup format. Backups using the old format can always be imported on newer versions of LXD.
+By default version `2.0` is used if the flag is not specified and the server has support for the `backup_version` API extension.
```
```suggestion
If the flag is not specified and the server has support for the `backup_version` API extension, version `2.0` is used by default.
```

roosterfish replied:
Fixed.

minaelee commented on lxd/api_internal.go:
```diff
@@ -663,28 +663,33 @@ func internalImportFromBackup(s *state.State, projectName string, instName strin
-	if instName != backupConf.Container.Name {
-		return fmt.Errorf("Instance name requested %q doesn't match instance name in backup config %q", instName, backupConf.Container.Name)
+	if instName != backupConf.Instance.Name {
+		return fmt.Errorf("Instance name requested %q doesn't match instance name in backup config %q", instName, backupConf.Instance.Name)
```
```suggestion
		return fmt.Errorf("Requested instance name %q doesn't match instance name %q in backup config ", instName, backupConf.Instance.Name)
```

roosterfish replied:
Fixed.

minaelee commented on lxd/api_internal.go:
```diff
@@ -740,25 +745,30 @@ func internalImportFromBackup(s *state.State, projectName string, instName strin
-		if dbVolume.Type != backupConf.Volume.Type {
-			return fmt.Errorf(`The type %q of the storage volume is not identical to the instance's type %q`, dbVolume.Type, backupConf.Volume.Type)
+		if dbVolume.Name != rootVol.Name {
+			return fmt.Errorf(`The name %q of the storage volume is not identical to the instance's name "%s"`, dbVolume.Name, backupConf.Instance.Name)
```
```suggestion
			return fmt.Errorf(`The storage volume name %q does not match the instance's name "%s"`, dbVolume.Name, backupConf.Instance.Name)
```

roosterfish replied:
Fixed.

minaelee commented on lxd/api_internal.go:
```diff
@@ -740,25 +745,30 @@ func internalImportFromBackup(s *state.State, projectName string, instName strin
 
-		err := s.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		if dbVolume.Type != rootVol.Type {
+			return fmt.Errorf(`The type %q of the storage volume is not identical to the instance's type %q`, dbVolume.Type, rootVol.Type)
```
```suggestion
			return fmt.Errorf(`The storage volume type %q does not match the instance's type %q`, dbVolume.Type, rootVol.Type)
```

roosterfish replied:
Fixed.

minaelee commented on lxd/backup/backup_config_utils.go:
```diff
@@ -81,6 +81,65 @@ func ConfigToInstanceDBArgs(state *state.State, c *config.Config, projectName st
+
+// DowngradeConfigFile changes from the new to the old metadata file format.
+// It's a noop in case the config is already using the old format.
+// Downgrading looses the information about any additional custom storage volumes
```
```suggestion
// Downgrading loses the information about any additional custom storage volumes
```

roosterfish replied:
Fixed.

minaelee commented on lxc/storage_volume.go:
```diff
@@ -2710,6 +2713,21 @@ func (c *cmdStorageVolumeExport) run(cmd *cobra.Command, args []string) error {
+	backupVersionSupported := d.HasExtension("backup_version")
+	backupVersion := api.BackupVersion(c.flagVersion)
+
+	// If the server supports setting the backup version set the selection version
```
```suggestion
	// If the server supports setting the backup version, set the selection version
```

roosterfish replied:
Fixed.

roosterfish commented:
The new linter errors should disappear after https://github.com/canonical/lxd/pull/15258 gets merged.

simondeziel commented on shared/version/api.go:
```diff
@@ -443,6 +443,7 @@ var APIExtensions = []string{
 	"client_cert_presence",
 	"clustering_groups_used_by",
 	"container_bpf_delegation",
+	"backup_version",
```
IIRC, such commit is best kept as the first in a series to make Tom's backport effort easier.

roosterfish replied:
Thanks for the hint, moved to the head now.

roosterfish replied:
Actually I moved it back now right before the changes to the public API (structs). Because all the commits before that should be allowed to be backported whilst the rest shouldn't (starting from the API extension commits).
So it's as you suggested but separating the "to-be-backported" commits from the ones starting with the API extension.

simondeziel replied:
Sounds good, thanks!

roosterfish replied:
Changed it again after realizing they are best suited at the beginning of the PR as the commits before wouldn't make sense as standalone as I have planned :)

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -4,13 +4,32 @@ import (
 )
 
+// VolumeConfig represents the config of a volume including its snapshots.
+type VolumeConfig struct {
```
"Config" here stands out compared to the other struct type names in the `Config` struct below.

As its all "config", the repeat of "Config" here seems unnecessary and inconsistent, what do you think?

roosterfish replied:
Agreed and fixed.

tomponline commented on lxd/backup/backup_config_utils.go:
```diff
@@ -81,6 +81,65 @@ func ConfigToInstanceDBArgs(state *state.State, c *config.Config, projectName st
+
+	// Set the corresponding backup format version if not set.
+	if backupConf.Version == 0 {
+		backupConf.Version = api.BackupMetadataVersion2
```
version 2? so we have no version 1, only 0 and 2?

tomponline replied:
or is version 0 equivalent to version 1 from a UX perspective?

roosterfish replied:
> or is version 0 equivalent to version 1 from a UX perspective?

Yep.

0 is just the default for uint which means if its not set then it must be the old format. The `.Version` field in the backup file won't ever be set to 1 as version 1 doesn't yet populate this field.

tomponline commented on lxd/instance_backup.go:
```diff
@@ -328,6 +329,15 @@ func instanceBackupsPost(d *Daemon, r *http.Request) response.Response {
+	// sending this field.
+	if req.Version == 0 {
+		req.Version = config.FileMetadataVersion
+	} else if !shared.ValueInSlice(req.Version, []api.BackupMetadataVersion{api.BackupMetadataVersion1, api.BackupMetadataVersion2}) {
```
Can we use `req.Version >  config.FileMetadataVersion` here?

roosterfish replied:
Fixed.

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -6,6 +6,9 @@ import (
 )
 
+// FileMetadataVersion represents the current version of the format used when writing a backup's metadata to file.
+const FileMetadataVersion = api.BackupMetadataVersion1
```
Is it used for migration as well, if so we should drop the `File` prefix.

roosterfish replied:
Fixed.

tomponline commented on lxd/storage/backend_lxd.go:
```diff
@@ -7068,6 +7063,17 @@ func (b *lxdBackend) UpdateInstanceBackupFile(inst instance.Instance, snapshots
+		// Downgrade the config in case the old backup format was requested.
+		// Do this as one of the last steps to allow working on the latest format beforehand.
+		if version == api.BackupMetadataVersion1 {
+			backup.DowngradeConfigFile(config)
```
I wonder if we should have a `backup.ConvertFormat(version uint)` function instead of upgrade/downgrade (also still not enjoying the `File` suffix here as we dont know its from a file).

Then the `ConvertFormat` can contain the logic to switch between the desired versions, rather than needing if statements before each call.

tomponline replied:
yes lets stick to uint32.

roosterfish replied:
Fixed, now using `uint32` everywhere.

tomponline commented on shared/api/storage_pool_volume_backup.go:
```diff
@@ -56,6 +56,12 @@ type StoragePoolVolumeBackupsPost struct {
+	// Example: 1
+	//
+	// API extension: backup_metadata_version
+	Version BackupMetadataVersion
```
I'm thinking we should stick with uint as data type rather than its own named type, as its pretty clear what a version is here.

roosterfish replied:
Fixed.

tomponline commented on doc/howto/storage_backup_volume.md:
```diff
@@ -139,6 +139,11 @@ You can add any of the following flags to the command:
+
+`--export-version`
+: If you intend to import the backup to an older version of LXD, set the version to `1` which will use the original (old)
+backup metadata format. Backups using the old format can always be imported on newer versions of LXD.
```
do we need a newline here?

tomponline replied:
I think we should split lines on  "Backup using..."

roosterfish replied:
Good catch, thanks!

roosterfish replied:
Fixed.

tomponline commented on doc/howto/storage_backup_volume.md:
```diff
@@ -139,6 +139,11 @@ You can add any of the following flags to the command:
+`--export-version`
+: If you intend to import the backup to an older version of LXD, set the version to `1` which will use the original (old)
+backup metadata format. Backups using the old format can always be imported on newer versions of LXD.
+If the flag is not specified and the server has support for the `backup_metadata_version` API extension, version `2` is used by default.
```
do you think we should add the latest supported backup version to the output of `/1.0`?

tomponline replied:
or will any future changes be via API extensions, like `backup_metadata_version3`?

roosterfish replied:
I like the idea of adding a list of supported versions to the `/1.0`. 
This way a user doesn't need to search for API extensions but can rather consult this list if `backup_metadata_version` extension is present.

roosterfish replied:
Fixed, as discussed its now represented by a range which is also tested in the added test suite.

tomponline reviewed (changes requested):
Thanks @roosterfish this is looking great.

There's a few ordering issues and a few minor questions/suggestions.

Thanks

tomponline commented:
@roosterfish does this PR deliver the roadmap item for improvements to `lxd recover` for restoring custom volumes?

tomponline reviewed (changes requested):
Please can you include some tests for the old and new formats too. Ta

roosterfish commented:
> @roosterfish does this PR deliver the roadmap item for improvements to `lxd recover` for restoring custom volumes?

Not yet. After this one we have to add custom storage vols to the now available list and then we can enhance `lxd recover`.

> Please can you include some tests for the old and new formats too. Ta

This is in the lxd-ci repo, see https://github.com/canonical/lxd-ci/pull/443

roosterfish commented on lxd/storage/backend_lxd.go:
```diff
@@ -774,30 +774,42 @@ func (b *lxdBackend) CreateInstanceFromBackup(srcBackup backup.Info, srcData io.
+		}
+	}
+
+	if rootVol != nil {
```
Tbh I just tried to keep the old logic but have made some tests now but in all cases (also using --empty) there is a root vol. 
So the only case where this might not be present is if somebody exports an instance, modifies the contents and then imports it. 
But in this case we can error out as the format seems corrupted.

roosterfish replied:
Fixed.

tomponline commented on lxc/export.go:
```diff
@@ -41,6 +43,8 @@ func (c *cmdExport) command() *cobra.Command {
 		i18n.G("Use storage driver optimized format (can only be restored on a similar pool)"))
 	cmd.Flags().StringVar(&c.flagCompressionAlgorithm, "compression", "", i18n.G("Compression algorithm to use (none for uncompressed)")+"``")
+	cmd.Flags().UintVar(&c.flagExportVersion, "export-version", 0,
+		i18n.G("Use a different metadata format version to be able to import the backup on older LXD versions")+"``")
```
You could make the flag a string type, defaulting to empty string, and then convert to a number internally before submitting to the API, with 0 being use the latest server version. This would avoid indicating `0` is a valid version on the CLI.

roosterfish replied:
Fixed as discussed.

roosterfish commented on lxd/backup/backup_info.go:
```diff
@@ -130,5 +130,8 @@ func GetInfo(r io.ReadSeeker, sysOS *sys.OS, outputPath string) (*Info, error) {
 	}
 
+	// Upgrade the config file in any case to the new format.
+	UpgradeConfigFile(result.Config)
```
I have squashed it into another (earlier) commit which introduced the upgrade/downgrade and already made use of the funcs.

roosterfish replied:
Fixed.

roosterfish commented on lxd/backup/config/backup_config.go:
```diff
@@ -33,3 +35,40 @@ type Config struct {
 }
+
+// RootVolumePool returns the pool of the root volume.
+// This is the first pool in the list of pools.
```
When upgrading from the version 1 format to version 2 the `Pool` field gets moved into `Pools` so it is the first element from the list.

tomponline replied:
It would be more proper to resolve the pool volume from the embedded profile and instance config in my view.

Is there a scenario where you see this wouldnt work?

roosterfish commented on lxd/backup/config/backup_config.go:
```diff
@@ -33,3 +35,40 @@ type Config struct {
+}
+
+// RootVolume returns the root volume.
+// This is the first volume in the list of volumes.
```
When upgrading from the version 1 format to version 2 the Volume field gets moved into Volumes so it is the first element from the list.
Can you think of a case where this might be different?

tomponline replied:
I'm not too keen on using positional elements to indicate special meaning, especially so when we have a pre-existing convention that is used through the rest of LXD.

roosterfish replied:
Makes sense. I have changed it accordingly. 

This also surfaced something else that was unintentionally working before. The same backup file format is used for both instances and custom storage vols. For the latter there always is only a single volume but having the function to return this vol for custom storage volumes be named `RootVolume()` feels odd.

Instead There is now a `RootVolume()` used when retrieving the actual root vol for instances and a `CustomVolume()` which returns the first and only volume in the list when retrieving the custom storage volume.

tomponline replied:
You could call it `PrimaryVolume()` if that helps for both?

roosterfish replied:
Fixed as discussed yesterday. Also the `CustomVolume()` func now checks if the file contains any traces that makes it look like its coming from an instance and errors out in this case.

roosterfish commented on lxd/backup/config/backup_config.go:
```diff
@@ -4,13 +4,32 @@ import (
+	Instance  *api.Instance           `yaml:"instance,omitempty"`
+	Snapshots []*api.InstanceSnapshot `yaml:"snapshots,omitempty"`
+	// Deprecated.
+	// Use Pools instead.
```
Fixed.

roosterfish commented on lxd/storage/backend_lxd.go:
```diff
@@ -6832,11 +6891,14 @@ func (b *lxdBackend) GenerateCustomVolumeBackupConfig(projectName string, volNam
 
-	return config, nil
+	return &backupConfig.Config{
+		Version: api.BackupMetadataVersion2,
```
Fixed. All constants commits are now moved to the beginning for clarity.

roosterfish commented on lxc/export.go:
```diff
@@ -76,6 +80,19 @@ func (c *cmdExport) run(cmd *cobra.Command, args []string) error {
+		if exportVersion != 0 {
+			req.Version = exportVersion
+		}
+	} else if !backupVersionSupported && !shared.ValueInSlice(exportVersion, []api.BackupMetadataVersion{api.BackupMetadataVersion1, 0}) {
```
Fixed.

roosterfish commented on lxd/backup.go:
```diff
@@ -255,6 +255,11 @@ func backupWriteIndex(sourceInst instance.Instance, pool storagePools.Pool, opti
 	}
 
+	// Downgrade the config in case the old backup format was requested.
+	if version == api.BackupMetadataVersion1 {
```
Fixed.

roosterfish commented on lxc/storage_volume.go:
```diff
@@ -2652,6 +2653,8 @@ func (c *cmdStorageVolumeExport) command() *cobra.Command {
 		i18n.G("Use storage driver optimized format (can only be restored on a similar pool)"))
 	cmd.Flags().StringVar(&c.flagCompressionAlgorithm, "compression", "", i18n.G("Define a compression algorithm: for backup or none")+"``")
+	cmd.Flags().UintVar(&c.flagExportVersion, "export-version", 0,
+		i18n.G("Use a different metadata format version to be able to import the backup on older LXD versions")+"``")
```
Fixed.
Also the last commit for 6.x adds a note that this flag can be used to produce backups which can be imported on older LXDs.

roosterfish commented on lxc/storage_volume.go:
```diff
@@ -2710,6 +2713,19 @@ func (c *cmdStorageVolumeExport) run(cmd *cobra.Command, args []string) error {
+		if exportVersion != 0 {
+			req.Version = exportVersion
+		}
+	} else if !backupVersionSupported && !shared.ValueInSlice(exportVersion, []api.BackupMetadataVersion{api.BackupMetadataVersion1, 0}) {
```
Fixed.

roosterfish commented:
@tomponline I have addressed all of the comments. You will find all of the requested changes in the first push (haven't rebased `main` in the middle to make reviewing easier).

There are a few more pushes after that for smaller tweaks in CI and two other ones when it comes to lookup of the instances root vol from the list of vols as it required special handling for instance snapshots. But I have added comments to the respective sections of the code.

tomponline commented on doc/api-extensions.md:
```diff
@@ -2642,3 +2642,8 @@ This adds a request option to set snapshot's target profile on instance copy to
+## `backup_metadata_version`
+
+Adds the field `version` when exporting instances and custom storage volumes to define the backup file format.
+In case the field is omitted, the original (old) version is used to maintain compatibility with older clients.
```
> In case the field is omitted, the original (old) version is used to maintain compatibility with older clients.

Is this still true, aren't we defaulting to the default preferred version that the server chooses if not specified?

Also, can you include some CLI usage examples here as this is useful for the docs. Ta

roosterfish replied:
True, I missed to update this sentence. Will also add some examples.

roosterfish replied:
Fixed.

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -4,6 +4,10 @@ import (
 
+// MetadataVersion represents the current version of the format used when writing a backup's metadata.
+// The metadata is used both for exporting backups and for migration.
+const MetadataVersion = api.BackupMetadataVersion1
```
Shall we call this DefaultMetadataVersion, and document it as the version used when the client doesnt specify a preferred version?

tomponline replied:
Presumably we will also need a MaxMetadataVersion constant too?

roosterfish replied:
The `MaxMetadataVersion` made me thinking. Remember this comment https://github.com/canonical/lxd/pull/15129#discussion_r2034931624. 
By using `req.Version >  config.FileMetadataVersion` instead, the API now rejects any version higher than the default which is not right for LXD 5.21 as it will support version 2 so we should have this max constant or the check that was in there previously.

But as the check is basically duplicated for instances and custom storage vols, I'll go with the const. Thanks.
I'll also split the test case so that there is something to be "backportable" to 5.21 too that would catch this error.

roosterfish replied:
I think we can add the "min" equivalent in the future when starting to deprecate version 1.

roosterfish replied:
The test as it is right now is actually fine and can be backported. Nvm.

roosterfish replied:
Fixed.

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -8,13 +8,28 @@ import (
+type Volume struct {
+	// Make sure to have the embedded structs fields inline to avoid nesting.
+	api.StorageVolume `yaml:",inline"`
+	Snapshots         []*api.StorageVolumeSnapshot `yaml:"snapshots,omitempty"`
```
Usually we have a new line between embedded types and struct fields IIRC?

roosterfish replied:
I have found both in the code but I think it makes sense for readability.

roosterfish replied:
Fixed.

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -33,3 +35,116 @@ type Config struct {
+	}
+
+	for deviceName, deviceConfig := range c.Instance.ExpandedDevices {
+		if deviceName == "root" {
```
The root device doesnt have to be called "root".

Can we use `instancetype.GetRootDiskDevice(entry.Devices)` here or in `RootVolumePool()` below instead of this function?

roosterfish replied:
Fixed.

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -33,3 +35,116 @@ type Config struct {
+	// Snapshot instances don't have the Instance field populated.
+	// A snapshot is always represented by a single volume.
+	// Therefore reuse the same tooling as when retrieving a custom volume.
+	volume, _ := c.CustomVolume()
```
hrm i dont like this because the error returned here might mention "custom volumes"

roosterfish replied:
Fixed. I now went with a `primaryVolume` which is generic enough to be called by `CustomVolume` and `RootVolume` depending on the use case.

roosterfish commented on lxd/backup/backup_config_utils.go:
```diff
@@ -81,6 +81,67 @@ func ConfigToInstanceDBArgs(state *state.State, c *config.Config, projectName st
 
+// ConvertFormat converts a backup config's metadata file format between versions.
+// In case the requested format is already present it's a noop.
+func ConvertFormat(backupConf *config.Config, version uint32) {
```
Properly copying the `*config.Config` is tricky as its full of pointers. Instead why not returning a cleanup hook which can be used by the caller if wanted?

tomponline replied:
the data has already gone by then though

tomponline replied:
we do have shared.DeepCopy btw.

roosterfish replied:
I knew there was something regarding deep copy but I didn't remember anymore. Great, I'll put it in then.

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -11,13 +11,29 @@ const DefaultMetadataVersion = api.BackupMetadataVersion1
+	Version   uint32                  `yaml:"version,omitempty"`
+	Instance  *api.Instance           `yaml:"instance,omitempty"`
+	Snapshots []*api.InstanceSnapshot `yaml:"snapshots,omitempty"`
+	Pools     []*api.StoragePool      `yaml:"pools,omitempty"`
```
As I was looking through the code it occurred to me that whilst `Snapshots` makes sense as a slice as order is important, the `Pools` and `Volumes` fields do not have order. Would it make sense to make these maps, keyed on their respective pool or volume name? This would make loading a specific pool/volume for read/update easier.

tomponline replied:
oh now, ofc, we cant do that for volumes as they may have the same name in different pools. Ignore me.

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -11,13 +11,29 @@ const DefaultMetadataVersion = api.BackupMetadataVersion1
+	Pools     []*api.StoragePool      `yaml:"pools,omitempty"`
+	Profiles  []*api.Profile          `yaml:"profiles,omitempty"`
+	Volumes   []*Volume               `yaml:"volumes,omitempty"`
+	Bucket    *api.StorageBucket      `yaml:"bucket,omitempty"`
```
I wonder if we should also give the same treatment to `Bucket` for consistency as it stands out a bit now.

tomponline replied:
hrm although we do not attach buckets to instances, so probably not infact.

Still we should at least consider https://github.com/canonical/lxd/pull/13924 and ensure this new format is compatible with that.

roosterfish replied:
Makes sense to me. I guess what we should have is, like for volumes, a dedicated struct which embeds the bucket api type as well as snapshots.
But just don't use a list as for volumes as there really can only ever be a single bucket.
The list of volumes only makes sense as we use the same structure for both instances (with multiple vols) and custom storage vols (with a single vol).

So I would propose this:

```go
type Bucket struct {
    // Make sure to have the embedded structs fields inline to avoid nesting.
    *api.StorageBucket   `yaml:",inline"`

    Snapshots []*api.StorageBackupSnapshot `yaml:"snapshots,omitempty"` // If this is the name of the field after landing support for bucket snapshots
}

type Config struct {
    Volumes []*Volume // support instances and custom storage vols with snapshots
    Bucket *Bucket    // support a single bucket with snapshots
}

```

tomponline replied:
yeah i think this makes sense as it allows for back up of bucket snapshots with differing config to its underlying volume config.

roosterfish replied:
Fixed in https://github.com/canonical/lxd/pull/15430.

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -37,3 +40,130 @@ type Config struct {
+	}
+
+	// Second try fetching the single volume for snapshot instances.
+	// Snapshot instances don't have the Instance field populated.
```
> 	// Snapshot instances don't have the Instance field populated.

Why is this?

roosterfish replied:
Ultimately this was added as part of [this commit](https://github.com/canonical/lxd/commit/1fedc735327358ba07767a58373852cd492931f3#diff-029fdd0e1afdacb0e66630c0d7e813772122db9a7d036060902838db315d9c4dR4553).
I suspect the reason is to keep the amount of data stored in the backup file for snapshots low as the same data is already written in the backup file of the instance itself.

This surfaced when running the instance migration tests when trying to migrate (copy) an instance snapshot to another new instance.
In this case getting the root volume is easy as the snapshot only has a single volume which is the root volume.

tomponline replied:
ah so this is only within the context of migration though right?

tomponline replied:
you cant export just a snapshot?

roosterfish replied:
Yes, currently snapshots of both instances and custom storage vols cannot be exported separately.

tomponline commented on lxd/backup/config/backup_config.go:
```diff
@@ -37,3 +40,130 @@ type Config struct {
+
+// RootVolume returns an instance's root volume from the list of volumes.
+// The volume's name matches the one of the instance.
+func (c *Config) RootVolume() (*Volume, error) {
```
I'm a bit confused why this function doesn't use the same technique as `rootVolPoolName` does to figure out the instance's root volume? There probably needs to be some explanation about this.

roosterfish replied:
Regardless of the name of the device, the volume under `volumes:` (v2) and `volume:` (v1) in the backup file is named after the name of the instance. Therefore the lookup has to be less involved.

tomponline commented on lxc/export.go:
```diff
@@ -76,6 +81,37 @@ func (c *cmdExport) run(cmd *cobra.Command, args []string) error {
+		return fmt.Errorf(i18n.G("Invalid export version %q: %w"), c.flagExportVersion, err)
+	}
+
+	versionUint32 := uint32(versionUint)
```
this intermediate var is probably unnecessary, as you can perform the logic comparisons on versiobnUint first and then just do:

```go
req.Version = uint32(versionUint)
```

roosterfish replied:
Fixed in https://github.com/canonical/lxd/pull/15430.

tomponline commented on lxc/export.go:
```diff
@@ -76,6 +81,37 @@ func (c *cmdExport) run(cmd *cobra.Command, args []string) error {
+		if versionUint32 != 0 {
+			req.Version = versionUint32
+		}
+	} else if !backupVersionSupported && versionUint32 > api.BackupMetadataVersion1 {
```
although does it prevent comparison with the constant?

roosterfish replied:
Yeah I have added the variable so that I can both use it for assignment to `req.Version` and in this check.

roosterfish replied:
But I'll move the entire block into its own func as suggested in the other comment so we have it at a single place only.

roosterfish replied:
Fixed in https://github.com/canonical/lxd/pull/15430.

tomponline commented on lxd/storage/backend_lxd.go:
```diff
@@ -7074,6 +7069,18 @@ func (b *lxdBackend) UpdateInstanceBackupFile(inst instance.Instance, snapshots
 			return err
 		}
 
+		// Downgrade the config in case the old backup format was requested.
```
can we do this just before vol.MountTask so we dont go through the overhead of mounting the volume only to fail on yaml conversion?

roosterfish replied:
Fixed in https://github.com/canonical/lxd/pull/15430.

roosterfish replied:
Thanks for the suggestion, I'll try to keep this in mind for the future when doing refactoring as it's clearly an improvement.

tomponline commented:
LGTM, there's some minor questions/nits, the most interesting one for me was the statements around .Instance not being populated for instance snapshots.

roosterfish commented on lxc/storage_volume.go:
```diff
@@ -2710,6 +2713,37 @@ func (c *cmdStorageVolumeExport) run(cmd *cobra.Command, args []string) error {
 		CompressionAlgorithm: c.flagCompressionAlgorithm,
 	}
 
+	backupVersionSupported := d.HasExtension("backup_metadata_version")
```
Fixed in https://github.com/canonical/lxd/pull/15430.
